ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -307,16 +324,38 @@ void onErrorResponse(ReplicaId replicaId) {
   private void handleBody(InputStream payload, MessageMetadata messageMetadata)
       throws IOException, MessageFormatException {
     ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
-    // @todo use the encryption key for decryption.
-    if (operationResult == null) {
-      BlobInfo blobInfo = new BlobInfo(MessageFormatRecord.deserializeBlobProperties(payload),
-          MessageFormatRecord.deserializeUserMetadata(payload).array());
-      getOptions().ageAtAccessTracker.trackAgeAtAccess(blobInfo.getBlobProperties().getCreationTimeInMs());
-      operationResult = new GetBlobResultInternal(new GetBlobResult(blobInfo, null), null);
+    serverBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+    ByteBuffer userMetadata = MessageFormatRecord.deserializeUserMetadata(payload);
+    if (encryptionKey == null) {
+      // if blob is not encrypted, move the state to Complete
+      operationResult =
+          new GetBlobResultInternal(new GetBlobResult(new BlobInfo(serverBlobProperties, userMetadata.array()), null),
+              null);
     } else {
-      // If the successTarget is 1, this case will never get executed.
-      // If it is more than 1, then, different responses will have to be reconciled in some way. Here is where that
-      // would be done. Since the store is immutable, currently we handle this by ignoring subsequent responses.
+      // submit decrypt job
+      progressTracker.startDecryptionTracker();
+      logger.trace("Submitting decrypt job for {}", blobId);
+      long startTimeMs = System.currentTimeMillis();
+      cryptoJobHandler.submitJob(
+          new DecryptJob(blobId, encryptionKey.duplicate(), null, userMetadata, cryptoService, kms,
+              (DecryptJob.DecryptJobResult result, Exception exception) -> {
+                logger.trace("Handling decrypt job callback results for {}", blobId);
+                routerMetrics.contentDecryptTimeMs.update(System.currentTimeMillis() - startTimeMs);
+                if (exception == null) {
+                  logger.trace("Successfully updating decrypt job callback results for {}", blobId);
+                  operationResult = new GetBlobResultInternal(
+                      new GetBlobResult(new BlobInfo(serverBlobProperties, result.getDecryptedUserMetadata().array()),
+                          null), null);
+                  progressTracker.setDecryptionSuccess();
+                } else {
+                  logger.trace("Exception {} thrown on decryption for {}", exception, blobId);
+                  setOperationException(
+                      new RouterException("Exception thrown on decrypting the content for " + blobId, exception,
+                          RouterErrorCode.UnexpectedInternalError));
+                }
+                progressTracker.setDecryptionFailed();
```

Comment:
shouldn't this be inside the `else`? Do all tests succeed?

Reply:
I'm still curious how tests were succeeding when this bug was around - something should have caught it.

Reply:
I see why this was succeeding - and it looks like that hole has been fixed